src/egraph.rs

```diff
@@ -455,41 +483,70 @@ impl<L: Language, N: Analysis<L>> EGraph<L, N> {
     /// If a copy is in the egraph, then [`add`] simply returns the id of the
     /// eclass in which the enode was found.
     ///
-    /// Like [`union`](EGraph::union), this modifies the e-graph,
-    /// so you must call [`rebuild`](EGraph::rebuild) any query operations.
+    /// Like [`union`](EGraph::union), this modifies the e-graph.
     ///
     /// [`add`]: EGraph::add()
-    pub fn add(&mut self, mut enode: L) -> Id {
-        self.lookup(&mut enode).unwrap_or_else(|| {
-            let id = self.unionfind.make_set();
-            if let Some(explain) = &mut self.explain {
-                explain.add(enode.clone(), id, id);
+    pub fn add(&mut self, enode: L) -> Id {
+        let id = self.add_internal(enode);
+        self.find(id)
+    }
+
+    /// Adds an enode to the egraph and also returns the the enode's id (uncanonicalized).
+    fn add_internal(&mut self, mut enode: L) -> Id {
+        let original = enode.clone();
+        if let Some(existing_id) = self.lookup_internal(&mut enode) {
+            let id = self.find(existing_id);
+            // when explanations are enabled, we need a new representative for this expr
+            if let Some(explain) = self.explain.as_mut() {
+                if let Some(existing_explain) = explain.uncanon_memo.get(&original) {
+                    *existing_explain
+                } else {
+                    let new_id = self.unionfind.make_set();
+                    explain.add(original, new_id, new_id);
+                    self.unionfind.union(id, new_id);
+                    explain.union(existing_id, new_id, Justification::Congruence, true);
+                    new_id
+                }
+            } else {
+                existing_id
+            }
+        } else {
+            let id = self.make_new_eclass(enode);
+            if let Some(explain) = self.explain.as_mut() {
+                explain.add(original, id, id);
             }
-            log::trace!("  ...adding to {}", id);
-            let class = EClass {
-                id,
-                nodes: vec![enode.clone()],
-                data: N::make(self, &enode),
-                parents: Default::default(),
-            };
 
-            // add this enode to the parent lists of its children
-            enode.for_each(|child| {
-                let tup = (enode.clone(), id);
-                self[child].parents.push(tup);
-            });
+            // now that we updated explanations, run the analysis for the new eclass
+            N::modify(self, id);
```

Comment:
@mwillsey it is key that we do modify after all the explanations stuff, so this version is different than the one we discussed.